### PR TITLE
Luxianze/issue32

### PR DESCRIPTION
**Pull request by LUXIANZE**

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -1,20 +1,36 @@
...
+import { makeExecutableSchema } from "apollo-server";
 
 import resolvers from "./resolvers";
 import typeDefs from "./typeDefs";
+import { permissions } from "./utils/validator";
 
 dotenv.config();
 
+const schema = makeExecutableSchema({ typeDefs, resolvers });
+const scheaWithMiddleware = applyMiddleware(schema, permissions);
```
Comment (HowardYaw) [inline src/index.js]: typo? schemaMiddleware

File: src/resolvers/User.js [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+
+      return filteredUsers;
+    },
+    user: async (root, { id }, context, info) => {
+      /**
+       * - check null id
+       * - validate mongoose id
+       * - find by id
+       */
+      if (id) {
```
Comment (HowardYaw) [inline src/resolvers/User.js:64]: else will throw an exception?
or just return 200 with empty response

File: src/resolvers/User.js [hunk 3]
```diff
@@ -0,0 +1,279 @@
+import {
+  ApolloError,
+  AuthenticationError,
+  UserInputError,
+} from "apollo-server-express";
+import validator from "email-validator";
```
Comment (HowardYaw) [inline src/resolvers/User.js]: can rename the validator?
too much validate here I confuse
**not mandatory la**

File: src/typeDefs/User.js [hunk 4]
```diff
@@ -10,24 +10,32 @@ export default gql`
...
     password: String!
     dateOfBirth: String!
     contactNo: String!
     email: String!
-    ip: String!
-    vehicle: VehicleInput!
-    employeeType: String!
-    favouriteBranch: [FavouriteBranchInput!]!
+    ip: String
+    vehicle: [VehicleInput!]!
```
Comment (HowardYaw) [inline src/typeDefs/User.js]: mandatory field?
if employee or brand owner also need input vehicle?
you intent to save an empty array is it?

File: src/typeDefs/User.js [hunk 5]
```diff
@@ -66,4 +75,16 @@ export default gql`
...
+  enum UserType {
+    EMPLOYEE
+    CUSTOMER
+    BRANDOWNER
+  }
+
+  enum EmployeeType {
+    MANAGER
+    STAFF
+    NONE
```
Comment (HowardYaw) [inline src/typeDefs/User.js:88]: NONE meaning what?
if brand owner and customer show not have this field right?

File: src/typeDefs/root.js [hunk 6]
```diff
@@ -1,10 +1,11 @@
 import { gql } from "apollo-server-express";
-import { GraphQLScalarType } from 'graphql';
-import { Kind } from 'graphql/language';
+import { GraphQLScalarType } from "graphql";
+import { Kind } from "graphql/language";
 
 export default gql`
   scalar Date
-  
+  scalar Upload
```
Comment (HowardYaw) [inline src/typeDefs/root.js:5]: i guess this is for file upload
but we are using File?

**Comment by HowardYaw** [comment]
@LUXIANZE 
Few items suggest here:
- [x] gitignore package-lock.json
- [x] utils/index.js is empty should remove
- [x] UserInputValidator (validate signUp) purpose when there is a typeDefs

File: src/index.js (hunk 1 above)
Comment (LUXIANZE) [reply src/index.js]: Not typo, it is schema + middleware

File: src/typeDefs/User.js (hunk 4 above)
Comment (LUXIANZE) [reply src/typeDefs/User.js]: You're right. Will change.

File: src/resolvers/User.js (hunk 3 above)
Comment (LUXIANZE) [reply src/resolvers/User.js]: ok

File: src/typeDefs/User.js (hunk 5 above)
Comment (LUXIANZE) [reply src/typeDefs/User.js:88]: I cannot put null for enum. so use none to replace

File: src/typeDefs/root.js (hunk 6 above)
Comment (LUXIANZE) [reply src/typeDefs/root.js:5]: This is required when we use schema + middleware

File: src/resolvers/User.js (hunk 2 above)
Comment (LUXIANZE) [reply src/resolvers/User.js:64]: > @LUXIANZE
> Few items suggest here:
> 
> * [ ]  gitignore package-lock.json
> * [ ]  utils/index.js is empty should remove
> * [ ]  UserInputValidator (validate signUp) purpose when there is a typeDefs

- I think package-lock.json should remain as mentioned in [this post](https://stackoverflow.com/a/44210813)
- the other 2 I agree

File: src/typeDefs/root.js (hunk 6 above)
Comment (HowardYaw) [reply src/typeDefs/root.js:5]: ok

File: src/typeDefs/User.js (hunk 5 above)
Comment (HowardYaw) [reply src/typeDefs/User.js:88]: ok

**Pull request merged**